R17 updates: layout_stride

crtrott opened the pull request.

crtrott committed:
Make layout_left/right return const extents_type& from extents()

crtrott committed:
Bring layout_stride in line with R17

dalg24 commented on tests/test_layout_stride.cpp:
```diff
@@ -107,7 +107,8 @@ TEST(TestLayoutStrideListInitialization, test_list_initialization) {
   ASSERT_EQ(m.stride(1), 128);
-  ASSERT_EQ(m.strides(), (std::array<std::size_t, 2>{1, 128}));
+  ASSERT_EQ(m.strides()[0], 1);
+  ASSERT_EQ(m.strides()[1], 128);
```
Why did you change this?  Can we not compare collections?

crtrott replied:
nope it now returns span

dalg24 commented on include/experimental/__p0009_bits/submdspan.hpp:
```diff
@@ -397,7 +397,8 @@ constexpr auto _submdspan_impl(
   auto acc_pol = typename AP::offset_policy(src.accessor());
   return mdspan<
-    ET, decltype(map.extents()), typename decltype(_handled)::layout_type, decltype(acc_pol)
+    ET, remove_const_t<remove_reference_t<decltype(map.extents())>>,
+        typename decltype(_handled)::layout_type, remove_const_t<remove_reference_t<decltype(acc_pol)>>
```
Why is it not `remove_cv_t`?

crtrott replied:
we don't have volatile on this thing or?

crtrott committed:
Update default_accessor to R17

crtrott committed:
Some mdspan updates for R17

Most importantly the default constructor is disabled for mdspans of
dynamic_rank()==0

crtrott committed:
Revert change for returning span() for strides()

crtrott committed:
Harmonize mdspan with R17

crtrott committed:
Add explicit: thanks Mark!

crtrott committed:
Fix CPP14 build

crtrott committed:
Update README

mhoemmen commented on include/experimental/__p0009_bits/mdspan.hpp:
```diff
@@ -111,40 +115,78 @@ class mdspan
+       _MDSPAN_TRAIT(is_default_constructible, mapping_type) &&
+       _MDSPAN_TRAIT(is_default_constructible, accessor_type)
+     )
+   )
```
Does `= default` not work here?  I was wondering why you needed `MDSPAN_HAS_CXX_20`.

crtrott replied:
How do we attach sfinae requirements to implement the constraints on a non-tempalted default constructor without C++20 requires?

mhoemmen replied:
Ah, yes, that sounds hard.  It's probably best not to worry about it in the back-port.  Otherwise you would have to specialize the whole class just for the default constructor.

mhoemmen commented on include/experimental/__p0009_bits/mdspan.hpp:
```diff
@@ -111,40 +115,78 @@ class mdspan
   )
   MDSPAN_INLINE_FUNCTION
   explicit constexpr mdspan(pointer p, SizeTypes... dynamic_extents)
     // TODO @proposal-bug shouldn't I be allowed to do `move(p)` here?
```
You should be able to do `move(p)` here, as the accessor policy requirements say that `is_nothrow_move_constructible_v<A::pointer>` is `true`.

mhoemmen replied:
```suggestion
```
You should be able to do `move(p)` here, as the accessor policy requirements say that `is_nothrow_move_constructible_v<pointer>` is `true`.

mhoemmen commented on include/experimental/__p0009_bits/mdspan.hpp:
```diff
@@ -319,31 +395,26 @@ mdspan(ElementType*, SizeTypes...)
+  (_MDSPAN_TRAIT(is_pointer, std::remove_reference_t<Pointer>))
 )
-mdspan(const Pointer&) -> mdspan<std::remove_pointer_t<Pointer>, extents<size_t>>;
+mdspan(Pointer&&) -> mdspan<std::remove_pointer_t<std::remove_reference_t<Pointer>>, extents<size_t>>;
```
Oh wait, we wanted `const Pointer&` still, right?

crtrott replied:
yeah it doesn't matter really, Pointer&& does work here too. But I think people are getting worked up over stuff which doesn't matter. Since this is ONLY the deduction guide. I.e. what actually happens is determined by the constructor. This thing is only used to figure out template parameters and then a constructor is called ...

mhoemmen replied:
Yup, I agree -- I saw the e-mail list discussion.

crtrott committed:
Move pointers in mdspan constructors

mhoemmen commented on include/experimental/__p0009_bits/mdspan.hpp:
```diff
@@ -111,76 +115,113 @@ class mdspan
 
+#if !MDSPAN_HAS_CXX_20
   MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mdspan() = default;
+#else
```
Per previous discussion: Backporting the requirements to pre-C++20 would likely require duplicating and specializing the entire `mdspan` class.  That wouldn't be worth the benefit (since the class wouldn't compile anyway if the requirements weren't met).
